File: tests/sdk/go/s3_test.go
package sdk_test

import (
	"bytes"
	"context"
	"io"
	"strings"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/s3"
)

func TestS3BucketObjectRoundTrip(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	client := newS3(t, cfg)
	ctx := context.Background()

	bucket := strings.ToLower(uniquePrefix(t) + "-bucket")
	key := "hello.txt"
	body := []byte("noctaxris-sdk-s3")

	_, err := client.CreateBucket(ctx, &s3.CreateBucketInput{Bucket: aws.String(bucket)})
	if err != nil {
		t.Fatalf("CreateBucket: %v", err)
	}
	t.Cleanup(func() {
		_, _ = client.DeleteObject(ctx, &s3.DeleteObjectInput{Bucket: aws.String(bucket), Key: aws.String(key)})
		_, _ = client.DeleteBucket(ctx, &s3.DeleteBucketInput{Bucket: aws.String(bucket)})
	})

	list, err := client.ListBuckets(ctx, &s3.ListBucketsInput{})
	if err != nil {
		t.Fatalf("ListBuckets: %v", err)
	}
	found := false
	for _, b := range list.Buckets {
		if b.Name != nil && *b.Name == bucket {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("ListBuckets missing %s", bucket)
	}

	_, err = client.PutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
		Body:   bytes.NewReader(body),
	})
	if err != nil {
		t.Fatalf("PutObject: %v", err)
	}

	get, err := client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		t.Fatalf("GetObject: %v", err)
	}
	got, err := io.ReadAll(get.Body)
	_ = get.Body.Close()
	if err != nil {
		t.Fatalf("read GetObject body: %v", err)
	}
	if !bytes.Equal(got, body) {
		t.Fatalf("GetObject body=%q want=%q", got, body)
	}

	_, err = client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	if err != nil {
		t.Fatalf("DeleteObject: %v", err)
	}

	_, err = client.GetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	if err == nil {
		t.Fatal("expected GetObject after delete to fail")
	}
}

File: tests/sdk/nodejs/test/s3.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateBucketCommand,
  DeleteBucketCommand,
  DeleteObjectCommand,
  GetObjectCommand,
  ListBucketsCommand,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import { newS3, requireReady, uniquePrefix } from "../lib/helpers.mjs";

test("S3 bucket object round-trip", async (t) => {
  if (!(await requireReady(t))) return;
  const client = newS3();
  const bucket = `${uniquePrefix()}-bucket`.toLowerCase();
  const key = "hello.txt";
  const body = Buffer.from("noctaxris-sdk-s3");

  await client.send(new CreateBucketCommand({ Bucket: bucket }));
  t.after(async () => {
    try {
      await client.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }));
    } catch {
      /* ignore */
    }
    try {
      await client.send(new DeleteBucketCommand({ Bucket: bucket }));
    } catch {
      /* ignore */
    }
  });

  const list = await client.send(new ListBucketsCommand({}));
  const names = (list.Buckets || []).map((b) => b.Name);
  assert.ok(names.includes(bucket), `ListBuckets missing ${bucket}`);

  await client.send(
    new PutObjectCommand({ Bucket: bucket, Key: key, Body: body }),
  );
  const got = await client.send(new GetObjectCommand({ Bucket: bucket, Key: key }));
  const bytes = Buffer.from(await got.Body.transformToByteArray());
  assert.deepEqual(bytes, body);

  await client.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }));
  await assert.rejects(() =>
    client.send(new GetObjectCommand({ Bucket: bucket, Key: key })),
  );
});
